### ISSUE TITLE
traccar 4.4 error

### ISSUE THREAD

**Issue by vsatmydynipnet** [traccar/traccar#4297]
Hi,

i tried to update 4.3 to 4.4 as usual.
```
11.2-RELEASE-p9 FreeBSD 11.2-RELEASE-p9 #0: Tue Feb  5 15:30:36 UTC 2019     root@amd64-builder.daemonology.net:/usr/obj/usr/src/sys/GENERIC  amd64
```
Running different versions since a longer time, updates 4.0, 4.3 was no problems.

copied over conf/traccar.xml

but  on start it gets some db error:
```
2019-04-29 10:13:01  INFO: HikariPool-1 - Starting...
2019-04-29 10:13:01  WARN: Registered driver with driverClassName=com.mysql.jdbc.Driver was not found, trying direct instantiation.
2019-04-29 10:13:06 ERROR: HikariPool-1 - Exception during pool initialization. - Could not create connection to database server. Attempted reconnect 3 time
s. Giving up. - SQLNonTransientConnectionException (... < DataManager:139 < *:89 < Context:292 < Main:110 < ...)
2019-04-29 10:13:06 ERROR: Main method error - Failed to initialize pool: Could not create connection to database server. Attempted reconnect 3 times. Givin
g up. - PoolInitializationException (... < DataManager:139 < *:89 < Context:292 < Main:110 < ...)
```
i am not a java guy, any idea would be greatly appreciated.

switched back to 4.3 meanwhile, no stress.

**Comment by vsatmydynipnet** [comment]
Add: my sql setting:

    <--
    <entry key='database.driver'>com.mysql.jdbc.Driver</entry>
   -->
    <entry key='database.url'>jdbc:mysql://127.0.0.1:3306/MYDB?allowMultiQueries=true&amp;autoReconnect=true&amp;useUnicode=yes&amp;characterEncoding=UTF-8&amp;sessionVariables=sql_mode=ANSI_QUOTES</entry>
    <entry key='database.user'>MYUSER</entry>
    <entry key='database.password'>MYPASS</entry>

**Comment by tananaev** [comment]
1. Try to run manually and see what error you get.
2. Try to set `serverTimezone` in the JDBC string.

**Comment by vsatmydynipnet** [comment]
# startet from commandline with the correct user

same config works flawless with 4.3. i am no java guy, so kinda helpless :-)
``
$ cd /usr/local/www/traccar-4.4
$ /usr/local/bin/java -jar /usr/local/www/traccar-4.4/tracker-server.jar /usr/local/www/traccar-4.4/conf/traccar.xml
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/usr/local/www/traccar-4.4/lib/slf4j-jdk14-1.7.26.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/local/www/traccar-4.4/lib/logback-classic-1.2.3.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.JDK14LoggerFactory]
Exception in thread "main" java.lang.RuntimeException: java.lang.NullPointerException
	at org.traccar.Main.run(Main.java:152)
	at org.traccar.Main.main(Main.java:104)
Caused by: java.lang.NullPointerException
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:325)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at com.zaxxer.hikari.HikariConfig.attemptFromContextLoader(HikariConfig.java:899)
	at com.zaxxer.hikari.HikariConfig.setDriverClassName(HikariConfig.java:474)
	at org.traccar.database.DataManager.initDatabase(DataManager.java:124)
	at org.traccar.database.DataManager.<init>(DataManager.java:89)
	at org.traccar.Context.init(Context.java:292)
	at org.traccar.Main.run(Main.java:110)
	... 1 more
``

**Comment by vsatmydynipnet** [comment]
I changed my connection string to

    <entry key='database.url'>jdbc:mysql://127.0.0.1:3306/MYDB?allowMultiQueries=true&amp;autoReconnect=true&amp;useUnicode=yes&amp;character
Encoding=UTF-8&amp;sessionVariables=sql_mode=ANSI_QUOTE&amp;useLegacyDatetimeCode=false&amp;serverTimezone=UTC</entry>

but error persists

/usr/local/bin/java -jar /usr/local/www/traccar-4.4/tracker-server.jar /usr/local/www/traccar-4.4/conf/traccar.xml
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/usr/local/www/traccar-4.4/lib/slf4j-jdk14-1.7.26.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/local/www/traccar-4.4/lib/logback-classic-1.2.3.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.JDK14LoggerFactory]
Exception in thread "main" java.lang.RuntimeException: java.lang.NullPointerException
	at org.traccar.Main.run(Main.java:152)
	at org.traccar.Main.main(Main.java:104)
Caused by: java.lang.NullPointerException
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:325)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at com.zaxxer.hikari.HikariConfig.attemptFromContextLoader(HikariConfig.java:899)
	at com.zaxxer.hikari.HikariConfig.setDriverClassName(HikariConfig.java:474)
	at org.traccar.database.DataManager.initDatabase(DataManager.java:124)
	at org.traccar.database.DataManager.<init>(DataManager.java:89)
	at org.traccar.Context.init(Context.java:292)
	at org.traccar.Main.run(Main.java:110)
	... 1 more

**Comment by GoNzCiD** [comment]
I have the same error!
In 4.3 it works perfect.

My connection config is:
```
<entry key='database.driver'>com.mysql.jdbc.Driver</entry> 
<entry key='database.url'>jdbc:mysql://IP:PORT/traccar_database?useSSL=false&amp;allowMultiQueries=true&amp;autoReconnect=true&amp;useUnicode=yes&amp;characterEncoding=UTF-8&amp;sessionVariables=sql_mode=''</entry>
<entry key='database.user'>USER</entry> 
<entry key='database.password'>PASS</entry>
<entry key='database.positionsHistoryDays'>1000</entry>
```

Log:
```
2019-04-30 09:44:15  INFO: HikariPool-1 - Starting...
2019-04-30 09:44:15  WARN: Registered driver with driverClassName=com.mysql.jdbc.Driver was not found, trying direct instantiation.
2019-04-30 09:44:20 ERROR: HikariPool-1 - Exception during pool initialization. - Could not create connection to database server. Attempted reconnect 3 times. Giving up. - SQLNonTransientConnectionException (... < DataManager:139 < *:89 < Context:292 < Main:110 < ...)
```

**Comment by tananaev** [comment]
You can safely ignore that warning for now.

**Comment by vsatmydynipnet** [comment]
Here it is no warning, it simply exits without starting it. :-)

**Comment by tananaev** [comment]
Your last error is clearly different.

**Comment by Phantomatz** [comment]
Thanks, the hint with the timecode works for me. Adding 

&amp;useLegacyDatetimeCode=false&amp;serverTimezone=UTC 

did the job

@vsatmydynipnet, I think you lost the "S" of ANSI_QUOTES while implementing the string. Maybe this might cause your problem.
<entry key='database.url'>jdbc:mysql://127.0.0.1:3306/MYDB?allowMultiQueries=true&amp;autoReconnect=true&amp;useUnicode=yes&amp;character
Encoding=UTF-8&amp;sessionVariables=sql_mode=ANSI_QUOTE&amp;useLegacyDatetimeCode=false&amp;serverTimezone=UTC</entry>

**Comment by vsatmydynipnet** [comment]
Thank you, but the S is not the problem.

Tried 4.5 just now, with and without serverTimezone setting, both results in error again

2019-05-05 10:41:28 ERROR: Main method error - NullPointerException (... < DataManager:124 < *:89 < Context:292 < Main:110 < ...)
2019-05-05 11:17:24 ERROR: Main method error - NullPointerException (... < DataManager:124 < *:89 < Context:292 < Main:110 < ...)

Running manual:

/usr/local/bin/java -jar /usr/local/www/traccar-4.5/tracker-server.jar /usr/local/www/traccar-4.5/conf/traccar.xml
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/usr/local/www/traccar-4.5/lib/slf4j-jdk14-1.7.26.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/local/www/traccar-4.5/lib/logback-classic-1.2.3.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.JDK14LoggerFactory]
Exception in thread "main" java.lang.RuntimeException: java.lang.NullPointerException
	at org.traccar.Main.run(Main.java:152)
	at org.traccar.Main.main(Main.java:104)
Caused by: java.lang.NullPointerException
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:325)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at com.zaxxer.hikari.HikariConfig.attemptFromContextLoader(HikariConfig.java:899)
	at com.zaxxer.hikari.HikariConfig.setDriverClassName(HikariConfig.java:474)
	at org.traccar.database.DataManager.initDatabase(DataManager.java:124)
	at org.traccar.database.DataManager.<init>(DataManager.java:89)
	at org.traccar.Context.init(Context.java:292)
	at org.traccar.Main.run(Main.java:110)
	... 1 more

Switching back to 4.3 meanwhile.

**Comment by dieterch** [comment]
In traccar release 4.4 and later, the mysql driver was changed to 8.0.xxx

to connect mysql server, the entry in **traccar.xml** needs to be:

`<entry key='database.driver'>com.mysql.jc.jdbc.Driver</entry>
`

This change solved the issue for me.

**Comment by tananaev** [comment]
You can continue using old driver class, so that's not an issue.

**Comment by dieterch** [comment]
Thanks for the answer, 

in my case the old driver class was not found and the server crashed with an exception, the new one worked immediately.

Dieter

Von meinem iPad gesendet

**Comment by GoNzCiD** [comment]
> 
> 
> In traccar release 4.4 and later, the mysql driver was changed to 8.0.xxx
> 
> to connect mysql server, the entry in **traccar.xml** needs to be:
> 
> `<entry key='database.driver'>com.mysql.jc.jdbc.Driver</entry> `
> 
> This change solved the issue for me.

I've try in traccar v4.5 with this configuration and it throw an error:
`ERROR: Main method error - com.mysql.jc.jdbc.Driver - ClassNotFoundException (... < DataManager:120 < *:89 < Context:292 < Main:110 < ...)`

**Comment by tananaev** [comment]
It probably means that you haven't installed it correctly.

**Comment by GoNzCiD** [comment]
> 
> 
> It probably means that you haven't installed it correctly.

I'm running it in docker over alipe 3.8 and jre 8.202, and then run traccar by manual installation.
It's needed a newer java version?

**Comment by dieterch** [comment]
I was using Anton‘s Dockerfile from traccar/traccar-docker alpine.

then I manually changed the Version String to ‚4.5‘ within that file and ran docker build.

Docker start Parameters were according to the documentation on traccar.org, using a MySQL docker Container.

All software i used was pulled from the docker hub and the traccar reproduzier

Von meinem iPhone gesendet

**Comment by dieterch** [comment]
I mean Traccar Repository


Von meinem iPhone gesendet

**Comment by vsatmydynipnet** [comment]
> Thank you, but the S is not the problem.
> 
> Tried 4.5 just now, with and without serverTimezone setting, both results in error again
> 
> ```
> 2019-05-05 10:41:28 ERROR: Main method error - NullPointerException (... < DataManager:124 < *:89 < Context:292 < Main:110 < ...)
> 2019-05-05 11:17:24 ERROR: Main method error - NullPointerException (... < DataManager:124 < *:89 < Context:292 < Main:110 < ...)
> ```
> 
> Running manual:
> 
> ```
> /usr/local/bin/java -jar /usr/local/www/traccar-4.5/tracker-server.jar /usr/local/www/traccar-4.5/conf/traccar.xml
> SLF4J: Class path contains multiple SLF4J bindings.
> SLF4J: Found binding in [jar:file:/usr/local/www/traccar-4.5/lib/slf4j-jdk14-1.7.26.jar!/org/slf4j/impl/StaticLoggerBinder.class]
> SLF4J: Found binding in [jar:file:/usr/local/www/traccar-4.5/lib/logback-classic-1.2.3.jar!/org/slf4j/impl/StaticLoggerBinder.class]
> SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
> SLF4J: Actual binding is of type [org.slf4j.impl.JDK14LoggerFactory]
> Exception in thread "main" java.lang.RuntimeException: java.lang.NullPointerException
> 	at org.traccar.Main.run(Main.java:152)
> 	at org.traccar.Main.main(Main.java:104)
> Caused by: java.lang.NullPointerException
> 	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:325)
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
> 	at com.zaxxer.hikari.HikariConfig.attemptFromContextLoader(HikariConfig.java:899)
> 	at com.zaxxer.hikari.HikariConfig.setDriverClassName(HikariConfig.java:474)
> 	at org.traccar.database.DataManager.initDatabase(DataManager.java:124)
> 	at org.traccar.database.DataManager.<init>(DataManager.java:89)
> 	at org.traccar.Context.init(Context.java:292)
> 	at org.traccar.Main.run(Main.java:110)
> 	... 1 more
> ```
> 
> Switching back to 4.3 meanwhile.

Please I do not use docker an any box. The problem why iI opened the issue still exists.

**Comment by vsatmydynipnet** [comment]
[SOLVED]

It looks like there was another change witch caused the crash after update to 4.4, or 4.5. 

Changed the string finally to:

<entry key='database.url'>jdbc:mysql://127.0.0.1:3306/MYDB?serverTimezone=UTC&amp;useSSL=false&amp;allowMultiQueries=true&amp;autoReconnect=true&amp;useUnicode=yes&amp;characterEncoding=UTF-8&amp;sessionVariables=sql_mode=''</entry>

OLD was 

serverTimezone=UTC&amp;allowMultiQueries=true&amp;autoReconnect=true&amp;useUnicode=yes&amp;characterEncoding=UTF-8&amp;sessionVariables=sql_mode=ANSI_QUOTES

Assume it was the sql_mode parameter.

**Comment by GoNzCiD** [comment]
It has been solved to me with 4.5 too
